src/Evolution/Systems/ScalarTensor/Tags.hpp

```diff
@@ -70,67 +72,46 @@ struct ScalarMass : db::SimpleTag {
-    get(*result) = get(psi);
-  }
-  using base = CSW<CurvedScalarWave::Tags::Psi>;
-};
-
-/*!
- * \brief Compute tag for observing the evolved variable Pi.
- * \details This tag copies the data from the evolved variable Tag
- * CurvedScalarWave::Tags::Pi into the wrapped tag used for observation.
- */
-struct CSWPiCompute : CSW<CurvedScalarWave::Tags::Pi>, db::ComputeTag {
-  using argument_tags = tmpl::list<CurvedScalarWave::Tags::Pi>;
-  using return_type = Scalar<DataVector>;
-  static constexpr void function(const gsl::not_null<return_type*> result,
-                                    const Scalar<DataVector>& pi) {
-    get(*result) = get(pi);
-  }
-  using base = CSW<CurvedScalarWave::Tags::Pi>;
-};
-
-/*!
- * \brief Compute tag for observing the evolved variable Phi.
- * \details This tag copies the data from the evolved variable Tag
- * CurvedScalarWave::Tags::Phi into the wrapped tag used for observation.
+ * \brief Compute tag to retrieve the values of CurvedScalarWave variables
+ * with the same name as in the ::gh systems.
+ * \note Since we also add compute tags for these quantities, we do not make
+ * this a derived class of `Tag`. Otherwise, we would have tags with repeated
+ * base tags in the `ObservationBox`.
  */
-template <size_t Dim>
-struct CSWPhiCompute : CSW<CurvedScalarWave::Tags::Phi<Dim>>, db::ComputeTag {
-  using argument_tags = tmpl::list<CurvedScalarWave::Tags::Phi<Dim>>;
-  using return_type = tnsr::i<DataVector, Dim>;
+template <typename Tag>
+struct CswCompute : Csw<Tag>, db::ComputeTag {
+  using argument_tags = tmpl::list<Tag>;
+  using base = Csw<Tag>;
+  using return_type = typename base::type;
   static constexpr void function(const gsl::not_null<return_type*> result,
-                                    const tnsr::i<DataVector, Dim>& phi) {
-    for (size_t i = 0; i < Dim; ++i) {
-      result->get(i) = phi.get(i);
+                                 const return_type& csw_var) {
+    for (size_t i = 0; i < csw_var.size(); ++i) {
+      make_const_view(make_not_null(&std::as_const((*result)[i])), csw_var[i],
+                      0, csw_var[i].size());
     }
   }
-  using base = CSW<CurvedScalarWave::Tags::Phi<Dim>>;
 };
 
 /*!
  * \brief Computes the scalar-wave one-index constraint.
  * \details The one-index constraint is assigned to a wrapped tag to avoid
  * clashes with the ::gh constraints during observation.
+ * \note We do not use ScalarTensor::Tags::CswCompute to retrieve the
+ * CurvedScalarWave constraints since we do not want to add the bare compute
+ * tags (::CurvedScalarWave::Tags::OneIndexConstraintCompute and
+ * ::CurvedScalarWave::Tags::TwoIndexConstraintCompute) directly in the
+ * ObservationBox, since that will make them observable and would lead to a
+ * clash with the ::gh constraint tags.
  */
 template <size_t Dim>
-struct CSWOneIndexConstraintCompute
-    : CSW<CurvedScalarWave::Tags::OneIndexConstraint<Dim>>,
+struct CswOneIndexConstraintCompute
```

Comment:
Does the `CswCompute` tag not work on the constraints? Naively I would've expected it to work on any tag that holds some sort of tensor.

Reply:
It does work. But...
in the executable we would have
`using observe_fields = tmpl::list<..., CurvedScalarWave::Tags::TwoIndexConstraintCompute<volume_dim>, ScalarTensor::Tags::CswCompute<CurvedScalarWave::Tags::TwoIndexConstraint<volume_dim>>, ..., ::gh::Tags::TwoIndexConstraintCompute<volume_dim>, ...>`

So in principle someone could still write in the input file:
`VariablesToObserve:
- Csw(TwoIndexConstraint)
- TwoIndexConstraint`

and trigger an error when the code tries to write twice in the same file for the two index constraint where for the second variable it is still confused on which one to observe.

So I wanted to avoid adding at all `CurvedScalarWave::Tags::TwoIndexConstraintCompute<volume_dim>` in observe_fields.

At least that's I understand is happening

Reply:
Oh, I see... Hmm, I don't have a great solution for that at the moment, so doing what you are doing is fine :) Thanks!